cpp/src/gandiva/engine.cc

```diff
@@ -256,23 +289,26 @@ Status Engine::LoadPreCompiledIR() {
   /// Parse the IR module.
   llvm::Expected<std::unique_ptr<llvm::Module>> module_or_error =
       llvm::getOwningLazyBitcodeModule(std::move(buffer), *context());
-  if (!module_or_error) {
-    // NOTE: llvm::handleAllErrors() fails linking with RTTI-disabled LLVM builds
-    // (ARROW-5148)
-    std::string str;
-    llvm::raw_string_ostream stream(str);
-    stream << module_or_error.takeError();
-    return Status::CodeGenError(stream.str());
-  }
-  std::unique_ptr<llvm::Module> ir_module = std::move(module_or_error.get());
+  // NOTE: llvm::handleAllErrors() fails linking with RTTI-disabled LLVM builds
+  // (ARROW-5148)
+  ARROW_RETURN_NOT_OK(VerifyAndLinkModule(module_, std::move(module_or_error)));
+  return Status::OK();
+}
 
-  // set dataLayout
-  SetDataLayout(ir_module.get());
+llvm::MemoryBufferRef AsLLVMMemoryBuffer(
+    const std::unique_ptr<arrow::Buffer>& arrow_buffer) {
```

Comment:
Two things:
1) a simple `const Buffer&` seems sufficient here
2) please make sure that all private functions are made `static`, or moved into the anonymous namespace

Reply:
Fixed.